presto-main/src/main/java/io/prestosql/operator/GroupedTopNRowNumberAccumulator.java

```diff
@@ -0,0 +1,483 @@
+    {
+        groupIdToHeapBuffer.allocateGroupIfNeeded(groupId);
+
+        long heapRootNodeIndex = groupIdToHeapBuffer.getHeapRootNodeIndex(groupId);
+        if (heapRootNodeIndex == UNKNOWN_INDEX || calculateRootRowNumber(groupId) < topN) {
+            heapInsert(groupId, rowReference.extractRowId());
+            return true;
+        }
+        else if (rowReference.compareTo(rowComparator, heapNodeBuffer.getRowId(heapRootNodeIndex)) < 0) {
+            heapPopAndInsert(groupId, rowReference.extractRowId(), rowIdEvictionListener);
+            return true;
+        }
+        else {
+            return false;
+        }
+    }
+
+    /**
+     * Drain the contents of this accumulator to the provided output row ID buffer.
+     * <p>
+     * Rows will be presented in increasing rank order. Draining will not trigger any row eviction callbacks.
+     *
+     * @param groupId
+     * @param rowIdOutput
+     * @return number of rows deposited to the output buffer
+     */
+    public long drainTo(long groupId, LongBigArray rowIdOutput)
+    {
+        long heapSize = groupIdToHeapBuffer.getHeapSize(groupId);
+        rowIdOutput.ensureCapacity(heapSize);
+        // Heap is inverted to output order, so insert back to front
+        for (long i = heapSize - 1; i >= 0; i--) {
+            rowIdOutput.set(i, peekRootRowId(groupId));
+            // No eviction listener needed because this is an explicit caller directive to extract data
+            heapPop(groupId, null);
+        }
+        return heapSize;
+    }
+
+    private long peekRootRowId(long groupId)
+    {
+        long heapRootNodeIndex = groupIdToHeapBuffer.getHeapRootNodeIndex(groupId);
+        verify(heapRootNodeIndex != UNKNOWN_INDEX, "No root to peek");
+        return heapNodeBuffer.getRowId(heapRootNodeIndex);
+    }
+
+    private long getChildIndex(long heapNodeIndex, HeapTraversal.Child child)
+    {
+        return child == HeapTraversal.Child.LEFT
+                ? heapNodeBuffer.getLeftChildHeapIndex(heapNodeIndex)
+                : heapNodeBuffer.getRightChildHeapIndex(heapNodeIndex);
+    }
+
+    private void setChildIndex(long heapNodeIndex, HeapTraversal.Child child, long newChildIndex)
+    {
+        if (child == HeapTraversal.Child.LEFT) {
+            heapNodeBuffer.setLeftChildHeapIndex(heapNodeIndex, newChildIndex);
+        }
+        else {
+            heapNodeBuffer.setRightChildHeapIndex(heapNodeIndex, newChildIndex);
+        }
+    }
+
+    /**
+     * Pop the root node off the group ID's max heap.
+     *
+     * @param groupId
+     * @param contextEvictionListener optional callback for the root node that gets popped off
+     */
+    private void heapPop(long groupId, @Nullable LongConsumer contextEvictionListener)
+    {
+        long heapRootNodeIndex = groupIdToHeapBuffer.getHeapRootNodeIndex(groupId);
+        checkArgument(heapRootNodeIndex != UNKNOWN_INDEX, "Group ID has an empty heap");
```

Comment:
checkArgument because you are not allowed to call the heapPop() on an empty group, which indicates an invalid argument passed given the preconditions of this method. Versus verify in other places where local and class-wide invariants are violated. Is that the standard you guys are also using?